### PR TITLE
package sync- capture and compare the latest commit

### PR DESCRIPTION
**Pull request by Sreenivas-Kalluru**
package sync- capture and compare the latest commit
I have captured the commit id in a file (commit.txt) that gets updated as during the pipeline step

### CODE REVIEW

**Comment by Sreenivas-Kalluru** [comment]
@ras0219-msft  @JavierMatosD  Could you please review and let me know if any comments?

File: scripts/rebuild.sh [hunk 1]
```diff
@@ -5,7 +5,7 @@ set -e
 cd -P -- "$(dirname -- "${BASH_SOURCE[0]}")"
 if [ ! -e ../vcpkg ]
 then
-    vcpkg_commit=d765306b074717dea8dc1c4723e1b025acb61c2d
+    vcpkg_commit=b73601a4616c21923cfb6f3a5263a9788165efbd
```
[the remark this answers is not in the document]
Comment (Sreenivas-Kalluru) [reply scripts/rebuild.sh]: reading the updated commit hash from commit.txt is integrated into the #121

File: scripts/generateGitHubStars.js [hunk 2]
```diff
@@ -6,24 +6,30 @@ const { Octokit } = require('@octokit/rest');
...
-    return response.data.stargazers_count;
+        const response = await octokit.rest.repos.get({ owner, repo });
+        if (response.status != 200) {
+            console.log(`Failed to get stars for ${url}`);
+            return 0;
+        }
+
+        return response.data.stargazers_count;
+    } catch (error) {
+console.log(error);
```
[the remark this answers is not in the document]
Comment (Sreenivas-Kalluru) [reply scripts/generateGitHubStars.js]: Now returned the default value 0 to be consistent with the other error checks

File: .github/workflows/validate.yml [hunk 3]
```diff
@@ -21,6 +23,10 @@ jobs:
           key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
           restore-keys: ${{ runner.os }}-node-
 
+      - name: Sync Package metadata
+        run: scripts/syncMetadata.sh
+        if: ${{ env.Enable_SyncPackage == 'true' }}
+
```
[the remark this answers is not in the document]
Comment (Sreenivas-Kalluru) [reply .github/workflows/validate.yml:28]: created a new yaml file(.github/workflows/sync-package-periodically.yml) that triggers every Saturday at midnight and saves the latest commit hash in the commit.txt file

File: .github/workflows/sync-package-periodically.yml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  schedule:
+    - cron: '0 0 * * 6'
+jobs:
+  validate:
+    runs-on: ubuntu-latest
+    
+    env:
+     Codeql.Enabled: true
+     Enable_SyncPackage: false
+
```
Comment (JavierMatosD) [inline .github/workflows/sync-package-periodically.yml:13]: Where are these variables used?

File: .github/workflows/sync-package-periodically.yml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      - name: Rebuild and validate website
+        run: scripts/rebuild.sh
+
+      - name: Commit Results
+        run: |
+            git config --local user.email "actions@github.com"
+            git config --local user.name "GitHub Actions"
+            git add commit.txt
+            git commit -m 'Data Updated' || 'No changes to commit'
```
Comment (JavierMatosD) [inline .github/workflows/sync-package-periodically.yml:36]: The 'No changes to commit' is just a string. It won't execute if `git commit -m` fails. Did you mean `echo 'No changes to commit'`?

File: .github/workflows/sync-package-periodically.yml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+            git push origin || echo "No changes to commit"
+            
+      - name: Check for differences
+        run: |
+          if ! git diff --quiet; then
+            echo "::error ::Detected changes when regenerating website."
+            git diff --stat
+            exit 1
+          fi
```
Comment (JavierMatosD) [inline .github/workflows/sync-package-periodically.yml:45]: Should we check for differences before we commit?

File: scripts/generateGitHubStars.js [hunk 7]
```diff
@@ -6,24 +6,31 @@ const { Octokit } = require('@octokit/rest');
...
+
+        const response = await octokit.rest.repos.get({ owner, repo });
+        if (response.status != 200) {
+            console.log(`Failed to get stars for ${url}`);
+            return 0;
+        }
+
+        return response.data.stargazers_count;
+    } catch (error) {
+        console.log(error);
```
Comment (JavierMatosD) [inline scripts/generateGitHubStars.js:31]:
```suggestion
        console.log("Error fetching stars for repo.");
```

File: scripts/rebuild.sh [hunk 8]
```diff
@@ -3,9 +3,15 @@
...
 cd -P -- "$(dirname -- "${BASH_SOURCE[0]}")"
-if [ ! -e ../vcpkg ]
-then
-    vcpkg_commit=d765306b074717dea8dc1c4723e1b025acb61c2d
+if [ ! -e ../vcpkg ]; then
+    vcpkg_commit=d5b03c125afee1d9cef38f4cfa77e229400fb48a
+    filename="commit.txt"
+    if [ -f "$filename" ]; then
+        content=$(cat commit.txt)
+        if [ $vcpkg_commit != $content ]; then
```
Comment (JavierMatosD) [inline scripts/rebuild.sh:11]:
```suggestion
        if [ "$vcpkg_commit" != "$content" ]; then
```

**Review by JavierMatosD — CHANGES_REQUESTED** [review]
Looks good to me with the additional suggested changes.

File: .github/workflows/sync-package-periodically.yml (hunk 5 above)
Comment (Sreenivas-Kalluru) [reply .github/workflows/sync-package-periodically.yml:36]: Yes it is, echo 'No changes to commit' if git commit -m fails

File: .github/workflows/sync-package-periodically.yml (hunk 4 above)
Comment (Sreenivas-Kalluru) [reply .github/workflows/sync-package-periodically.yml:13]: Enable_SyncPackage is not in scope and removed it.  The Codeql.Enabled setting is required as part of Microsoft compliance check.

File: .github/workflows/sync-package-periodically.yml (hunk 5 above)
Comment (JavierMatosD) [reply .github/workflows/sync-package-periodically.yml:36]:
```suggestion
            git commit -m 'Data Updated' ||  echo 'No changes to commit'
```

File: .github/workflows/sync-package-periodically.yml (hunk 4 above)
Comment (JavierMatosD) [reply .github/workflows/sync-package-periodically.yml:13]:
```suggestion
     Codeql.Enabled: true
```
If it isn't being used we should remove it.

File: scripts/rebuild.sh [hunk 9]
```diff
@@ -3,9 +3,15 @@
...
-then
-    vcpkg_commit=d765306b074717dea8dc1c4723e1b025acb61c2d
+if [ ! -e ../vcpkg ]; then
+    vcpkg_commit=d5b03c125afee1d9cef38f4cfa77e229400fb48a
+    filename="commit.txt"
+    if [ -f "$filename" ]; then
+        content=$(cat commit.txt)
+        if [ "$vcpkg_commit" != "$content" ]; then
+            vcpkg_commit=$content
+        fi
```
Comment (ras0219-msft) [inline scripts/rebuild.sh:13]:
```suggestion
        vcpkg_commit=$content
```

File: scripts/rebuild.sh [hunk 10]
```diff
@@ -3,9 +3,15 @@
...
+    fi
```
Comment (ras0219-msft) [inline scripts/rebuild.sh:14]:
```suggestion
    vcpkg_commit=$(cat commit.txt)
    if [ ! -e commit.txt ]; then; exit 1; fi
```

I don't see any reason why we need to have a separate commit here.

File: scripts/syncMetadata.sh [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+content=$(cat commit.txt)
+
+if [ $vcpkg_commit != $content ]
+then
+    git init ../vcpkg
+    git -C ../vcpkg fetch --depth 1 https://github.com/Microsoft/vcpkg $vcpkg_commit
+    git -C ../vcpkg checkout FETCH_HEAD
+    echo $vcpkg_commit > commit.txt
+fi
```
Comment (ras0219-msft) [inline scripts/syncMetadata.sh:24]: Why is this not just:

```suggestion
$(git ls-remote $repo_url | head -n1 | awk '{ print $1 }') > commit.txt
```

File: .github/workflows/sync-package-periodically.yml [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+      - name: Rebuild and validate website
+        run: scripts/rebuild.sh
+
+      - name: Commit Results
+        run: |
+            git config --local user.email "actions@github.com"
+            git config --local user.name "GitHub Actions"
+            git add commit.txt
+            git commit -m 'Data Updated' || 'No changes to commit'
+            git push origin || echo "No changes to commit"
```
Comment (ras0219-msft) [inline .github/workflows/sync-package-periodically.yml:37]: Future hardening should separate regenerating the website (no write permissions) from pushing to origin (write permissions).

File: scripts/rebuild.sh (hunk 10 above)
Comment (Sreenivas-Kalluru) [reply scripts/rebuild.sh]: Incorporated the suggestions.

File: scripts/rebuild.sh (hunk 9 above)
Comment (Sreenivas-Kalluru) [reply scripts/rebuild.sh]: Incorporated the suggestion.

**Review by JavierMatosD — APPROVED** [review]

**Pull request merged**